src/hmm/topology.cc

```diff
@@ -239,148 +118,178 @@ void Topology::Check() {
   }
   return entries_[phone2idx_[phone]];
 }
 
 int32 Topology::NumPdfClasses(int32 phone) const {
   // will throw if phone not covered.
-  const TopologyEntry &entry = TopologyForPhone(phone);
-  int32 max_pdf_class = 0;
-  for (size_t i = 0; i < entry.size(); i++) {
-    max_pdf_class = std::max(max_pdf_class, entry[i].forward_pdf_class);
-    max_pdf_class = std::max(max_pdf_class, entry[i].self_loop_pdf_class);
+  const fst::StdVectorFst &entry = TopologyForPhone(phone);
+
+  std::set<int32> pdfs;
+  for (fst::StateIterator<fst::StdVectorFst> siter(entry); !siter.Done(); siter.Next()) {
+    StateId state_id = siter.Value();
+    for (fst::ArcIterator<fst::StdVectorFst> aiter(entry, state_id);
+         !aiter.Done(); aiter.Next()) {
+      pdfs.insert(aiter.Value().ilabel);
+    }
   }
-  return max_pdf_class+1;
+  return pdfs.size();
+}
+
+void Topology::GetPhoneToNumPdfClasses(std::vector<int32> *phone2num_pdf_classes) const {
+  KALDI_ASSERT(!phones_.empty());
+  phone2num_pdf_classes->clear();
+  phone2num_pdf_classes->resize(phones_.back() + 1, -1);
+  for (auto phone: phones_)
+    (*phone2num_pdf_classes)[phone] = NumPdfClasses(phone);
 }
 
 int32 Topology::MinLength(int32 phone) const {
-  const TopologyEntry &entry = TopologyForPhone(phone);
-  // min_length[state] gives the minimum length for sequences up to and
-  // including that state.
-  std::vector<int32> min_length(entry.size(),
-                                std::numeric_limits<int32>::max());
-  KALDI_ASSERT(!entry.empty());
-
-  min_length[0] = (entry[0].forward_pdf_class == -1 ? 0 : 1);
-  int32 num_states = min_length.size();
-  bool changed = true;
-  while (changed) {
-    changed = false;
-    for (int32 s = 0; s < num_states; s++) {
-      const HmmState &this_state = entry[s];
-      std::vector<std::pair<int32, BaseFloat> >::const_iterator
-          iter = this_state.transitions.begin(),
-          end = this_state.transitions.end();
-      for (; iter != end; ++iter) {
-        int32 next_state = iter->first;
-        KALDI_ASSERT(next_state < num_states);
-        int32 next_state_min_length = min_length[s] +
-            (entry[next_state].forward_pdf_class == -1 ? 0 : 1);
-        if (next_state_min_length < min_length[next_state]) {
-          min_length[next_state] = next_state_min_length;
-          if (next_state < s)
-            changed = true;
-          // the test of 'next_state < s' is an optimization for speed.
-        }
+  using Weight = typename fst::StdFst::Weight;
+  using StateId = typename fst::StdFst::StateId;
+  const fst::StdVectorFst& this_topo = TopologyForPhone(phone);
+  // 1) Prepare a new FST with arc weight of 1.f and final state weight of 0.f
+  // (Note that 0.f == Weight::One() in Tropical Semiring).
+  // Since we are using the Std
+  // We need to use a VectorFst in order to mutate members
+  std::unique_ptr<fst::StdVectorFst> topo_copy(this_topo.Copy());
```

Comment:
You just do
`StdVectorFst topo_copy(this_topo);`
the constructor does the same as Copy().